### PR TITLE
FIX trigger indptr/indices copy when data are copied in astype

### LINKED ISSUE #8678: csr_matrix.astype raise ValueError when indices are not sorted and indices or indptr are read-only
### Reproducing code example:
```py
import numpy as np

from scipy import sparse

data = [2, 1, 4, 3]
indices = [1, 0, 1, 0]
indptr = [0, 2, 4]
matrix = sparse.csr_matrix((data, indices, indptr))
print('matrix.todense():\n', repr(matrix.todense()))
# To trigger this bug you need to make sure that the indices are not sorted
print('matrix.has_sorted_indices:', matrix.has_sorted_indices)

# To trigger this bug you need to make sure that either .indptr or .indices are read-only
matrix.indices.flags.writeable = False
matrix.astype(np.float64)
```

### Error message:
```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/test.py in <module>()
     12 
     13 matrix.indices.flags.writeable = False
---> 14 matrix.astype(np.float64)

/home/local/lesteve/miniconda3/lib/python3.6/site-packages/scipy/sparse/data.py in astype(self, dtype, casting, copy)
     69         if self.dtype != dtype:
     70             return self._with_data(
---> 71                 self._deduped_data().astype(dtype, casting=casting, copy=copy),
     72                 copy=copy)
     73         elif copy:

/home/local/lesteve/miniconda3/lib/python3.6/site-packages/scipy/sparse/data.py in _deduped_data(self)
     32     def _deduped_data(self):
     33         if hasattr(self, 'sum_duplicates'):
---> 34             self.sum_duplicates()
     35         return self.data
     36 

/home/local/lesteve/miniconda3/lib/python3.6/site-packages/scipy/sparse/compressed.py in sum_duplicates(self)
   1007         if self.has_canonical_format:
   1008             return
-> 1009         self.sort_indices()
   1010 
   1011         M, N = self._swap(self.shape)

/home/local/lesteve/miniconda3/lib/python3.6/site-packages/scipy/sparse/compressed.py in sort_indices(self)
   1053         if not self.has_sorted_indices:
   1054             _sparsetools.csr_sort_indices(len(self.indptr) - 1, self.indptr,
-> 1055                                           self.indices, self.data)
   1056             self.has_sorted_indices = True
   1057 

ValueError: WRITEBACKIFCOPY base is read-only
```

### Scipy/Numpy/Python version information:
```
1.0.0 1.14.2 sys.version_info(major=3, minor=6, micro=4, releaselevel='final', serial=0)
```

For the record, I tried with the dev wheels we are using in the scikit-learn tests, with the same result:
```
1.1.0.dev0+895a774 1.15.0.dev0+e2cb4e1 sys.version_info(major=3, minor=6, micro=4, releaselevel='final', serial=0)
```

### Additional comments

Why this matters in a scikit-learn context: scikit-learn use joblib for parallel computing which encourages the use of read-only memmap to share data-structures between worker processes. In this case the CSR matrix `.indices` and `.indptr` are read-only memmaps. Here are a few scikit-learn issues where a similar error was encountered: https://github.com/scikit-learn/scikit-learn/issues/6614 and https://github.com/scikit-learn/scikit-learn/issues/10916.

Even when `.indices` and `.indptr` are writeable, using `csr_matrix.astype` may have a side effect on the CSR matrix by sorting its indices. This may not be a big problem in practice, but it still feels slightly surprising.

### PR DESCRIPTION
**Pull request by glemaitre**
#### Reference issue

closes gh-8678
closes gh-8679
supersede gh-8679

#### What does this implement/fix?

In some cases, it happens that the internal arrays of the scipy matrices are memmaped in read-only mode (e.g. using `joblib.Parallel`). Calling `astype` calls `sum_duplicates` (called within `_deduped_data`) that intends to sort the indices of the matrix in place. It will therefore raise an error due to the read-only property.

Here, I trigger a copy of `intptr` and `indices` in the case the `dtype` does not match.

#### Additional information

As reported in the issue, this problem was reported several times in scikit-learn: https://github.com/scikit-learn/scikit-learn/issues/15924, https://github.com/scikit-learn/scikit-learn/issues/25935, etc.

### CODE REVIEW

**Comment by glemaitre** [comment]
In terms of non-regression tests, the minimal example in the issue https://github.com/scipy/scipy/issues/8678 could be used. However, I assume that it would be best to test several formats of sparse matrices.

Unfortunately, I am a bit lost with the test suite. I think that it should go in `scipy/sparse/tests/test_base.py`. I see a test that is written for `astype` but since we need special conditions to trigger the regression (read-only + non-sorted indices), I am not sure what is the best way to go forward.

**Comment by perimosocordiae** [comment]
For testing, it should be sufficient to add another method next to the existing `test_astype` method, like this one:

```python
def test_astype_immutable(self):
    D = array([[2.0 + 3j, 0, 0],
               [0, 4.0 + 5j, 0],
               [0, 0, 0]])
    S = self.spmatrix(D)
    if hasattr(S, 'data'):
        S.data.flags.writeable = False
    for x in supported_dtypes:
        D_casted = D.astype(x)
        S_casted = S.astype(x)
        assert_equal(S_casted.dtype, D_casted.dtype)
```

I wonder if the root cause issue is actually in the implementation of `_deduped_data`, rather than `astype`. If we add support for handling immutable member arrays in `_deduped_data`, that should fix both this and potentially other lurking bugs.

**Comment by glemaitre** [comment]
> I wonder if the root cause issue is actually in the implementation of _deduped_data, rather than astype. If we add support for handling immutable member arrays in _deduped_data, that should fix both this and potentially other lurking bugs.

It was what you intended originally in https://github.com/scipy/scipy/pull/8679. Some of the failures in the CI seem to be that the `_deduped_data` return a `data` array that is not any more "aligned" with the `indptr` and `indices`. For instance, `sum_duplicates` will sort the indices. So it means that we cannot use `self._with_data` that try to keep the original indices since they are not synced with the data returned by `_deduped_data`.

If we make changes in `_deduped_data`, it seems that we would need to return the associated indices.

**Comment by glemaitre** [comment]
The current error is raise when a LIL matrix call `to_csr` and intends to flatten the array using the `lil_flatten_to_array` function. The Cython signature does not contain `const` for the buffer while in the case of read-only, we need to add this qualifier.

File: scipy/sparse/tests/test_base.py
Comment (glemaitre) [inline scipy/sparse/tests/test_base.py:1304]: Here we only define the `data` to be read-only. However, from the original issue, we could also have read-only `intptr` and `indices`.

**Review by tylerjereddy — COMMENTED** [review]
This patch allows the full test suite to pass for me locally on this branch, though it doesn't win any prizes for elegance. Just providing in case it helps get this unstuck.

```diff
diff --git a/scipy/sparse/_csparsetools.pyx.in b/scipy/sparse/_csparsetools.pyx.in
index 5c47852c0..e5972eac0 100644
--- a/scipy/sparse/_csparsetools.pyx.in
+++ b/scipy/sparse/_csparsetools.pyx.in
@@ -176,7 +176,12 @@ def _lil_get_lengths_{{NAME}}(object[:] input,
 
 {{define_dispatch_map('_LIL_GET_LENGTHS_DISPATCH', '_lil_get_lengths', IDX_TYPES)}}
 
-def lil_flatten_to_array(object[:] input,
+ctypedef fused obj_fused:
+    object
+    double
+
+
+def lil_flatten_to_array(const obj_fused[:] input,
                          cnp.ndarray output):
     return _LIL_FLATTEN_TO_ARRAY_DISPATCH[output.dtype](input, output)
 
@@ -311,7 +316,7 @@ def _lil_fancy_set_{{PYIDX}}_{{PYVALUE}}(cnp.npy_intp M, cnp.npy_intp N,
 
 
 def lil_get_row_ranges(cnp.npy_intp M, cnp.npy_intp N,
-                       object[:] rows, object[:] datas,
+                       const obj_fused[:] rows, const obj_fused[:] datas,
                        object[:] new_rows, object[:] new_datas,
                        object irows,
                        cnp.npy_intp j_start,

```

File: scipy/sparse/_csparsetools.pyx.in
Comment (perimosocordiae) [inline scipy/sparse/_csparsetools.pyx.in:179]: This is unfortunate, but based on https://github.com/cython/cython/issues/2485 it seems like the best workaround for now.

Let's add a comment mentioning the Cython limitation and link to that issue, so that when/if it gets resolved we can remove the hack.

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Overall this is looking fine. Working around the Cython limitation isn't so bad, especially if we can remove the fused type hack after https://github.com/cython/cython/pull/4712 is merged.

Eventually I'll need to clean up the `_deduped_data()` situation, but I don't want to block this PR on that.

**Comment by perimosocordiae** [comment]
The one failing test (mac/py3.11) has a strange error message:

> meson.build:1:0: ERROR: Cython compiler 'cython' cannot compile programs

It doesn't seem to be related to this PR, though, so I'm +1 to merge now.

**Comment by glemaitre** [comment]
Shall I restart the build or do we need to wait for another review for the PR to be merged?

**Comment by perimosocordiae** [comment]
Let's wait a few more days to see if someone else wants to review, and if not I'll merge it then.

**Comment by perimosocordiae** [comment]
No more comments, so I'll merge now.

**Pull request merged**
